ImputationPipeline/ScoreBGE/ScoreBGE.py

```diff
@@ -0,0 +1,266 @@
+import pandas as pd
+import numpy as np
+import pysam
+
+class BGEScorer():
+    def __init__(self, ref_dict_path, prs_weights_path):
+        """
+        Initialize the BGEScorer object.
+
+        Args:
+            ref_dict_path (str): The file path to the reference dictionary.
+            prs_weights_path (str): The file path to the PRS weights.
+        """
+        self.ref_dict = self._read_ref_dict(ref_dict_path)
+        self.prs_weights = self._read_prs_weights(prs_weights_path)
+        self.sample_names = None
+
+        self.gvcf_sample_score = None
+        self.gvcf_sites_scored = None
+        self.gvcf_sites_scored_set = None
+        self.gvcf_low_quality_sites = None
+
+        self.vcf_sample_score = None
+        self.vcf_sites_scored = None
+        self.vcf_num_sites_not_found = None
+
+    def _read_ref_dict(self, ref_dict_path):
+        with open(ref_dict_path) as ref_dict:
+            return [line.split('\t')[1].replace('SN:', '') for line in ref_dict if line.startswith('@SQ')]
+
+    def _read_prs_weights(self, prs_weights_path):
+        prs_weights = pd.read_table(prs_weights_path, dtype={'effect_allele': str, 'weight': np.float64, 'chr': str, 'bp': np.int32, 'ref': str, 'alt': str})
+
+        prs_weights[['contig', 'position', 'ref', 'alt']] = prs_weights['CHR:BP:REF:ALT'].str.split(':', expand=True)
+        prs_weights = prs_weights.drop(['CHR:BP:REF:ALT'], axis=1)
+        prs_weights['locus'] = prs_weights['contig'] + ':' + prs_weights['position']
+
+        prs_weights = prs_weights.astype({'effect_allele': str, 'weight': np.float64, 'contig': str, 'position': np.int32, 'ref': str, 'alt': str})
+
+        return prs_weights.sort_values(by='position').sort_values(kind='stable', by='contig', key=lambda contig: contig.map(self.ref_dict.index))
+    
+    def _compare_record_and_weight(self, record, weight):
+        if record.contig != weight.contig:
+            return self.ref_dict.index(record.contig) - self.ref_dict.index(weight.contig)
+        if record.pos <= weight.position <= record.stop:
+            return 0
+        return record.pos - weight.position
+
+    def _check_biallelic(self, record):
+        if record is not None and len(record.alts) > 1:
+            raise RuntimeError('Only biallelic variants are supported: ' + str(record))
```

Comment:
Basically because of the `DS` dosage field. According to the header line
```
##FORMAT=<ID=DS,Number=1,Type=Float,Description="Genotype dosage">
```
it's a Number=1 field, which means that there is one entry per line. Some online research suggests that some tools may emit it as Number=A, which would mean one entry per allele, but I haven't found any clear standard of how that value would then be interpreted. Given that in this pipeline we only ever have biallelic imputed sites I figured it's not worth dealing with it. Let me know if you disagree @kachulis

Reply:
got it, ok yeah that seems reasonable.